### ISSUE TITLE
Problem running fabio on OpenBSD

### ISSUE THREAD

**Issue by stephane-martin** [fabiolb/fabio#265]
Hello,

I compiled fabio on OpenBSD using go 1.8.0.

Fabio runs, but doesn't answer requests. A simple curl to the listening port does not return.

The VM only has one core, so that may be a problem. Setting GOMACPROCS (in env or in fabio conf) doesn't solve the problem though.

**Comment by magiconair** [comment]
Did you try the OpenBSD binary from the 1.4.2 release?

https://github.com/fabiolb/fabio/releases/tag/v1.4.2

`GOMAXPROCS` only affects how many OS threads are created. Since Go has a cooperative scheduler this has no effect on the number of requests it can handle. On a single core system `GOMAXPROCS` will be set to 1. However, I don't think that this is the problem. I'm no OpenBSD expert. Is this a 32 or 64 bit system?

**Comment by stephane-martin** [comment]
It's a 64bits VM (at vultr.com), with OpenBSD 6.1.

I tried to recompile fabio 1.4.3 with go 1.8. This time it works properly.
I also tried the official release from github. It works properly too :)

One thing strikes me though:
- on your binary ldd fabio gives "not a dynamic executable"
- on the binary from my compilation ldd gives:
```
fabio-gopath/bin/fabio
/usr/lib/libpthread.so.22.0
/usr/lib/libc.so.88.0
/usr/libexec/ld.so
```
How do you avoid the shared libraries dependencies ?

**Comment by magiconair** [comment]
Try with -tags "netgo"

Frank Schröder

**Comment by stephane-martin** [comment]
Correction: i was stupidly pointing curl to the fabio instance on a linux box. I did retry to run fabio 1.4.3 on OpenBSD, and it still doesn't answer HTTP requests at all.

**Comment by stephane-martin** [comment]
(thanks for the netgo tag. for reference: https://golang.org/doc/go1.5)

**Comment by stephane-martin** [comment]
I understood part of the problem. In main.go there are lines in func startServers like: `go proxy.ListenAndServe...`

In fact the ListenAndServe... functions can fail and return and error. But that error stays invisible because it is swallowed by the goroutine go statement. Would be good to catch those errors and log them.

On the OpenBSD box the network operations are indeed failing with error "protocol not available". I don't understand why ,yet. Maybe related to IPv6.

**Comment by magiconair** [comment]
Good catch. I'll fix that. 

Frank Schröder

**Comment by magiconair** [comment]
Pushed a change for catching the errors. As for the listener you might want to try `tcp6` or `tcp4` here to see whether that makes a difference:

https://github.com/fabiolb/fabio/blob/master/proxy/listen.go#L19

**Comment by stephane-martin** [comment]
Not really, got "[FATAL] ui: set tcp4 127.0.0.1:9998->127.0.0.1:XXX: protocol not available" instead of "[FATAL] ui: set tcp 127.0.0.1:9998->127.0.0.1:YYY: protocol not available" (here trying to query the UI, but same on the proxy port)

**Comment by stephane-martin** [comment]
Commenting `ln = tcpKeepAliveListener{ln.(*net.TCPListener)}` removes the problem. Not sure why...

**Comment by leprechau** [comment]
So it seems as though golang isn't properly telling OpenBSD how to configure the socket.  The `protocol not available` error is a raw OpenBSD `errno` (http://man.openbsd.org/errno.2).  These are wrapped by the golang syscall package (https://golang.org/src/syscall/zerrors_openbsd_amd64.go).

Is it possibly related to this bug: https://golang.org/src/net/ipsock.go?s=233:577#L3 ??

**Comment by stephane-martin** [comment]
I don't think it's related, anyway I disabled IPv6 on the interface during the bug hunt. Anyway I do have a workaround now, and the bug is most probably not in fabio source code. So don't bother too much.

**Comment by leprechau** [comment]
Disabling keepalive probably isn't a good solution for a production workload.

**Comment by stephane-martin** [comment]
Further workaround: http://man.openbsd.org/sysctl.3
`(net.inet.tcp.always_keepalive) Act as if the option SO_KEEPALIVE was set on all TCP sockets.`

**Comment by leprechau** [comment]
Nice, gotta love OpenBSD for those network tunables.
